packages/pigeon/lib/java_generator.dart

```diff
@@ -413,7 +413,8 @@ void generateJava(JavaOptions options, Root root, StringSink sink) {
             String toWriteValue = '';
             if (!hostDatatype.isBuiltin &&
                 rootClassNameSet.contains(field.dataType)) {
-              toWriteValue = '${field.name}.toMap()';
+              final String fieldName = field.name;
+              toWriteValue = '($fieldName == null) ? null : $fieldName.toMap()';
```

Comment:
It looks like you are relying on the Dart stringification of `null` being correct for Java code, which seems confusing; `($fieldName == null) ? 'null' : '$fieldName.toMap()';` seems much clearer to me as intentionally outputing the word "null" for Java code.

Reply:
The null check is suppose to be happening at runtime, not at compilation time.  That's why it's written this way.  The code you suggest is executing the null check at compilation time.

Reply:
I completely missed that this was Java code, not `${}`'d Dart code. Maybe add an explicit comment here (`// Don't attempt to convert null Java values to maps`) to help other readers avoid the same mistake since it looks so much like Dart code.

Reply:
Yea, the color coding of github didn't help.  It's harder to miss with a proper editor:
![Screen Shot 2021-08-05 at 1 02 06 PM](https://user-images.githubusercontent.com/30870216/128413702-b7a00734-074c-46f0-8357-2001fa41df69.png)